### PR TITLE
[SPARK-30226][SQL] Remove withXXX functions in WriteBuilder

### PR DESCRIPTION
**Pull request by edrevo**
### What changes were proposed in this pull request?
Adding a `LogicalWriteInfo` interface as suggested by @cloud-fan in https://github.com/apache/spark/pull/25990#issuecomment-555132991 

### Why are the changes needed?
It provides compile-time guarantees where we previously had none, which will make it harder to introduce bugs in the future.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Compiles and passes tests

### CODE REVIEW

**Comment by edrevo** [comment]
Do I need to open a new JIRA for this @cloud-fan ?

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
@edrevo yes, please create a new JIRA

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.write;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * This interface contains write information that data sources can use when generating a
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java]: This interface describes the logical information of the input data that is being written to the data source.

We can update the doc of `PhysicalWriteInfo` accordingly.

Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java]: `logical write information`

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+   * running at the same time, or a query is restarted and resumed. {@link BatchWrite} can use
+   * this id to identify the query.
+   */
+  String queryId();
+
+  /**
+   * the schema of the input data from Spark to data source.
+   */
+  StructType schema();
+}
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java]: How about we put `CaseInsensitiveStringMap` in it as well?

**Comment by cloud-fan** [comment]
@edrevo any updates?

**Comment by edrevo** [comment]
Sorry for the delay @cloud-fan , I'll pick this up again this week

**Comment by edrevo** [comment]
@cloud-fan , I have addressed your PR feedback.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/StagedTable.java [hunk 3]
```diff
@@ -32,10 +33,11 @@
  * {@link StagingTableCatalog#stageCreate(Identifier, StructType, Transform[], Map)} or
  * {@link StagingTableCatalog#stageReplace(Identifier, StructType, Transform[], Map)} to prepare the
  * table for being written to. This table should usually implement {@link SupportsWrite}. A new
- * writer will be constructed via {@link SupportsWrite#newWriteBuilder(CaseInsensitiveStringMap)},
- * and the write will be committed. The job concludes with a call to {@link #commitStagedChanges()},
- * at which point implementations are expected to commit the table's metadata into the metastore
- * along with the data that was written by the writes from the write builder this table created.
+ * writer will be constructed via
+ * {@link SupportsWrite#newWriteBuilder(CaseInsensitiveStringMap, LogicalWriteInfo)}, and the write
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/StagedTable.java]: this needs update.

Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/StagedTable.java]: should be `newWriteBuilder(LogicalWriteInfo)`

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/StagingTableCatalog.java [hunk 4]
```diff
@@ -39,9 +40,9 @@
  * TABLE AS SELECT operation, if the catalog does not implement this trait, the planner will first
  * drop the table via {@link TableCatalog#dropTable(Identifier)}, then create the table via
  * {@link TableCatalog#createTable(Identifier, StructType, Transform[], Map)}, and then perform
- * the write via {@link SupportsWrite#newWriteBuilder(CaseInsensitiveStringMap)}. However, if the
- * write operation fails, the catalog will have already dropped the table, and the planner cannot
- * roll back the dropping of the table.
+ * the write via {@link SupportsWrite#newWriteBuilder(CaseInsensitiveStringMap, LogicalWriteInfo)}.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/StagingTableCatalog.java]: ditto

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  /**
+   * the schema of the input data from Spark to data source.
+   */
+  StructType schema();
+
+  /**
+   * the options that the user specified when writing the dataset
+   */
+  CaseInsensitiveStringMap options();
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java]: nit: can we move options to the first?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVTable.scala [hunk 6]
```diff
@@ -49,8 +49,9 @@ case class CSVTable(
     CSVDataSource(parsedOptions).inferSchema(sparkSession, files, parsedOptions)
   }
 
-  override def newWriteBuilder(options: CaseInsensitiveStringMap): WriteBuilder =
-    new CSVWriteBuilder(options, paths, formatName, supportsDataType)
+  override def newWriteBuilder(
+      info: LogicalWriteInfo): WriteBuilder =
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVTable.scala]: nit: this can be in one line.

**Review by cloud-fan — APPROVED** [review]
LGTM except a few minor comments

**Comment by cloud-fan** [comment]
ping @edrevo

**Comment by edrevo** [comment]
@cloud-fan, latest PR feedback addressed

**Comment by cloud-fan** [comment]
LGTM, also cc @rdblue @brkyvz

**Comment by rdblue** [comment]
I don't see a compelling reason to change the API again this close to a release. Why does this need to happen? What is the benefit? The only one I see is minor and in my opinion not worth the extra churn.

**Comment by cloud-fan** [comment]
I've posted the reason a long time ago: https://github.com/apache/spark/pull/26591#issuecomment-557360726

I think it's not minor and is an important change. We haven't cut branch yet so it's not too late.

**Comment by rdblue** [comment]
I think there are other ways to ensure you're using a builder correctly. For one thing, adding validations to the internal implementations makes sense.

In general, we need to stop making breaking changes to the API simply because it is marginally better.

**Comment by cloud-fan** [comment]
I think an argument here is if it's marginally better or much better. I think it's much better.

Another question is: when should we stop making breaking changes to V2? We have made so many API changes for DS v2 in Spark 3.0 and I'm confused why we can't make API changes now? If there is a point that we should stop making breaking changes, it should be the branch cut. Before that, I think it's worth to make API changes even if it's slightly better. We don't have chances to fix these small API problems anymore after branch cut. 

cc @srowen @dongjoon-hyun @dbtsai

**Comment by cloud-fan** [comment]
There is a proposal to do branch cut on Jan 31, so we still have more than a month to add new features. Since we all agree that this is an improvement(minor or significant), shall we get this in?

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
It has been over 2 weeks and no more objections so far. Since this is an improvement (type-safe is better than manual check) and DS v2 API change is not a concern before branch cut, I'm merging it to master. Please keep leaving comments if you have different opinions. Thanks!

**Pull request closed without merging**